### ISSUE TITLE
V2: Train, val, test split bugs

### ISSUE THREAD

**Issue by KnathanM** [chemprop/chemprop#645]
I've been thinking about #639 and found that the whole issue of splitting data into train, val, and test splits is complicated by letting users give multiple sources of data. I found that most possible splits users may want are not possible currently. Which of the possible cases for a random split do we want to allow (see below)? First though some background.  

### Is validation or testing optional?
| Dataset | Optional in Chemprop? | Optional in Astartes? |
| - | - | - |
| Train | Not optional for obvious reasons | Not optional for obvious reasons |
| Validation | No, fitting the trainer [takes](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/cli/train.py#L761) `val_loader`. |  [Yes](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/data/splitting.py#L81). |
| Test |  [Yes](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/cli/train.py#L763) in Chemprop. | [No](https://github.com/JacksonBurns/astartes/blob/main/astartes/main.py#L385). |

There is a disconnect between Chemprop and Astartes about what datasets are optional. We can work around this by just treating astartes `train_val_test_split()` as a three way data split and `train_test_split()` as a two way data split. But I would add a comment to the code to explain that something like `train_data, _, val_data = split_component(..., sizes=(0.8, 0, 0.2))` is not an error for `train_data, _, test_data = split_component(..., sizes=(0.8, 0, 0.2))`, even though the return of `split_component()` is `train, val, test`.

### Possible cases for random split
Users can give Chemprop 1-3 sources of data. Chemprop uses these sources to create datasets for training, validation, and testing. I'll focus just on a random split, the analysis below might not apply to other split types.
| Case | Source types | Desired datasets | Possible? (Y/N) |
| ---- | ------------ | ---------------- | --------------- |
| 1 | `--data-path` | Train | :x: No, if `--separate-val-path` and `--separate-test-path` are not given, `split_component()` is called which calls astartes which throws an [error](https://github.com/JacksonBurns/astartes/blob/main/astartes/main.py#L381) if you give it split sizes `(1, 0, 0)` because it wants to put some data in a test set. |
| 2 | `--data-path` | Val | :heavy_check_mark: No, train dataset can't be empty. Use the predition capability if you want to check how well an existing model does. |
| 3 | `--data-path` | Test | :heavy_check_mark: No, train dataset can't be empty. Use the predition capability if you want to check how well an existing model does. |
| 4 | `--data-path` | Train, Val | :x: No, [astartes](https://github.com/JacksonBurns/astartes/blob/main/astartes/main.py#L302) will throw an error if the test dataset is empty |
| 5 | `--data-path` | Train, Test | :x: No, Chemprop requires a validation dataset |
| 6 | `--data-path` | Val, Test | :heavy_check_mark: No, train dataset can't be empty. Use the predition capability if you want to check how well an existing model does. |
| 7 | `--data-path` | Train, Val, Test | :heavy_check_mark: Yes, this is the simple base case Chemprop defaults to. |
| 8 | `--data-path` and `--separate-val-path` | Train, Val | :x: No, #639 is about this. IMO there is no reason to not just assign the data-path data to the train dataset and the separate-val-path to the val dataset.| 
| 9 | `--data-path` and `--separate-val-path` | Train, Val, Test | No, for the same reason as 8. But also I'm not sure if it makes sense to have one set of data to split for training and testing and another for validation. That feels susceptible to data leakage to me. |
| 10 | `--data-path` and `--separate-test-path` | Train, Test | :x: No, Chemprop requires a validation dataset |
| 11 | `--data-path` and `--separate-test-path` | Train, Val, Test | :x: Yes, but the test portion of data-path data will be [thrown away](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/cli/train.py#L555) and the user can't set the test split size to be 0. |
| 12 | `--data-path`, `--separate-val-path`, and `--separate-test-path` | Train, Val, Test | :heavy_check_mark: Yes, this doesn't call astartes and just assigns the data to the datasets. |

### Chemprop throws away some data when a separate val or test path is given.
Should Chemprop throw a warning if the user gives a separate val or test path, but also gives split sizes that would reserve some of the data-path data for the val or test dataset? I think this should either throw an error (because 0.8 0.1 0.1 is the silent default even when a separate val or test path is given) or change the default split sizes when those extra data paths are given.

**Comment by JacksonBurns** [comment]
I'll just add a couple things here:
 - Ultimately the names don't matter, so `astartes` supports both 2-way and 3-way splits. What we _do_ with the two sets during a 2-way split doesn't matter, so I think `astartes` could support only train/test or only train/val if we wrap it appropriately. The no-way (train only) split is another issue.
 - I think Chemprop could easily support no-validation runs - if you pass an empty dataloader, and then in the `validation_step` function (current v2 implementation is [here](https://github.com/chemprop/chemprop/blob/v2/dev/chemprop/models/model.py#L156) simply `return` it should effectively skip it.
 - My partial takeaway from this very thorough table (thank you!) is that we are being too flexible in effort to support too many different use cases. I think that if users want to specify _any_ split separately, they should specify _all_ the splits separately - this could be consolidated into an argument like `--split-paths` which accepts some number of file paths.

There's a greater philosophical question here about if should really allow people to do anything other than a 3-way split. Anything without a validation set is prone to overfitting on the training data (and there would be no way of checking) and anything without a test can't report a _fair_ accuracy metric without data leakage.

**Comment by davidegraff** [comment]
I agree with @JacksonBurns's last point. V1 was highly flexible and supported many features (like these) the cost of the internal code-base being almost illegible and a real mess of branching logic. While we should strive to maximize compatibility of the two versions' respective CLIs, we should also ask if a particular feature makes sense to reimplement. Per this issue, I can't really think of a compelling case wherein a user would supply a separate validation path but construct their training and testing sets from the same source. With the caveat of early stoping, both training and validation are the same (in that they don't contribute to training). IMO, if you have a "special" validation set (i.e., you've supplied one,) then you should also have a "special" testing set. This was my original logic when I first scaffolded out the training CLI.

As to @JacksonBurns's suggestion of a `--split-paths` arg, we could also think about restructuring `--input` into a `{1..3}`-tuple (i.e., `parser.add_argument("--input", ..., nargs="+")` and validate `1 <= len(args.input) <= 3`). Then later in the code:
1. when we read in the data, we return a 3-tuple of train/val/test data:
```python
match input_path:
    case [p_train]:
        train_data = load_data(p_train)
        val_data = None
        test_data = None
    case [p_train, p_val]:
        train_data = load_data(p_train)
        val_data = load_data(p_val)
        test_data = None
    case [p_train, p_val, p_test]:
        train_data = load_data(p_train)
        val_data = load_data(p_val)
        test_data = load_data(p_test)
    case _:
        raise ValueError(f"Must supply between 1 and 3 arguments to `input_paths`! got: {len(input_paths)}")

return train_data, val_data, test_data
```
2. when we split the data, we accept in that 3-tuple and split depending on the null-ity of the inputs:
```python
match (train, val, test):
    case [train, None, None]:
        all_data = train
        # split back into train/val/test
    case [train, None, test]:
        all_data = train
        # split back into train/val
    case [train, val, test]:
        # do nothing
    case _:
        raise ValueError(f"Must supply some combination of {train}, {train, test}, or {train, val, test}!")
```
---
Re: the general point about lack of support for no validation set. Mechanically, we can achieve this as long as we allow for logic that ensures `val_loader` is `None`. If so, then Lightning won't execute the validation loop during training, and we don't have to alter any core model code.

**Comment by KnathanM** [comment]
Another note is that if the user is giving their own splits (`--separate-val-path` and `--separate-test-path`) and is also giving their own features/descriptors, they must specify the path to the same features/descriptors for the val and test sets (`--separate-val-atom-descriptors-path` for example). I wonder if it would be simpler to not allow separate val and test data paths and instead require that all data are in one file, all atom features are in one file, all bond features are in one file, etc. and then ask the user to give a file with the indices to use for train/val/test. The user could just vstack their train/val/test files and then give indices that are like train: 1-1000, val: 1001-1100, test: 1101-1200.

**Comment by kevingreenman** [comment]
As discussed in our meetings, I agree with the points above about v1 being too flexible and overcomplicating our codebase. 

@KnathanM I fully support your suggestion of only having a single argument for each of these input files and requiring a file with the split indices if a user wants to do the splits on their own, as long as we clearly document what format we expect for this indices file.

**Comment by davidegraff** [comment]
What about adding a split_column to the input dataset? If the argument is specified, then we use that column in the dataset map a datapoint to a split. The column should
- contain one of “train,” “val,” or “test”
- be possibly sparse. If one or more splits aren’t specified, then they will be assigned using the unspecified points, I.e., those with an empty string in the column.
- It will be expected that any split that is specified will not be assigned any additional points. If all three splits are assigned points and there are unassigned points remaining, they will **not** be used

using this, we can simplify the additional features files into a single one that is parallel w/r/t the input file. Thoughts?

**Comment by JacksonBurns** [comment]
My initial reaction is that I think writing the loading logic for that (data + splits in same file) with our current object model would not make sense. We do data loading/preparation and splitting entirely separately, so this might require re-writing the loading somewhat? I am probably lacking in creativity here, but I think the separate file is probably easier.

**Comment by davidegraff** [comment]
The current model builds splits and loads data in one function, [`build_splits`](https://github.com/chemprop/chemprop/blob/fc036fb9fb3b7d9eda72341d719cc27eb4df0a1f/chemprop/cli/train.py#L521), which (1) loads in an input file and then (2) splits it into train/val/test depending on the values of `separate_{val,test}_path`. It doesn't seem like the proposal is at odds with this model. The changes required would be (a) returning the split specification from `build_data_from_files` and (b) adjusting the rest of the function accordingly. Per my suggestion above, the split specification is a `list[str | None]`, parallel to the input data (i.e., a row in the CSV) containing one of: `"train"`, `"val"`, `"test"`, or `None`. Perhaps even easier is that with the addition of this column, we could rely on pandas `groupby` to do the dataset partitioning for us as well.

**Comment by KnathanM** [comment]
One potential issue with including the split_column in the input data file is it is less flexible for the case of specifying your own folds for cross validation. There can be a default cross validation that doesn't require specifying any splits, but potentially the user could want to use the same test set for all folds. I think having a separate yaml or toml file for specifying indices gives more flexibility. 

In terms of combining the inputs we also have the SMILES, the targets, the atom features, the bond features, atom descriptors, molecule descriptors, and I might have missed some. Currently we combine the SMILES and targets in one file while the rest are in their own parallel files. I think we combine the SMILES and targets in one file because they are both always required. The rest of the files are optional. I think having the other inputs (extra features, etc.) in separate files is better because then the user doesn't have to specify which columns correspond to atom features (potentially many), they just give the path to a separate optional file.

**Comment by KnathanM** [comment]
In our meeting today, we decided to try implementing having just one file for the SMILES, targets, extra features, etc. This can also include a split column for the simple case of not specifying your own cross validation folds. I'll update this again with any problems this implementation has.
